include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1324,6 +1324,292 @@ namespace Differentiation
+     *       std::vector<Tensor<2, dim, ADNumberType>> Grad_u(
+     *         n_q_points, Tensor<2, dim, ADNumberType>());
+     *       std::vector<Tensor<1, dim, ADNumberType>> Grad_msp(
+     *         n_q_points, Tensor<1, dim, ADNumberType>());
+     *       fe_values[u_fe].get_function_gradients_from_local_dof_values(
+     *         dof_values_ad, Grad_u);
+     *       fe_values[msp_fe].get_function_gradients_from_local_dof_values(
+     *         dof_values_ad, Grad_msp);
+     *
+     *       // This variable stores the cell residual vector contributions.
+     *       // IMPORTANT: Note that each entry is hand-initialized with a value
+     *       // of zero. This is a highly recommended practise, as some AD
+     *       // numbers appear not to safely initialize their internal data
+     *       // structures.
+     *       std::vector<ADNumberType> residual_ad (
+     *         n_dependent_variables, ADNumberType(0.0));
+     *
+     *       // Compute the cell total residual
+     *       //   = (internal + external) contributions
+     *       for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+     *       {
+     *         // Calculate the deformation gradient and magnetic field at this
+     *         // quadrature point
+     *         const Tensor<2, dim, ADNumberType> F =
+     *           unit_symmetric_tensor<dim>() + Grad_u[q_point];
+     *         const Tensor<1, dim, ADNumberType> H = -Grad_msp[q_point];
+     *         Assert(numbers::value_is_greater_than(determinant(F), 0.0),
+     *                ExcMessage("Negative jacobian detected!"));
+     *
+     *         // Extract some configuration dependent variables from our
+     *         // nonlinear constitutive law for the current quadrature point.
+     *         // In this way they only have to be computed once per quadrature
+     *         // point.
+     *         const SymmetricTensor<2,dim,ad_type> S = get_S(F,H);
+     *         const Tensor<1,dim,ad_type>          B = get_B(F,H);
+     *
+     *         // Define some position-dependent aliases, to make the assembly
+     *         // process easier to follow.
+     *         const double &JxW = fe_values.JxW(q_point);
+     *
+     *         // Add contribution of the internal forces:
+     *         // Note that we assemble the residual contribution directly
+     *         // as it is this vector that is to be automatically linearised.
+     *         for (unsigned int I = 0; I < n_dofs_per_cell; ++I)
+     *         {
+     *           // Determine the component and block associated with
+     *           // the I'th local degree-of-freedom.
+     *           const unsigned int block_I =
+     *             fe.system_to_base_index(I).first.first;
+     *
+     *           if (block_I == u_block) // u-terms
+     *           {
+     *             // Variation of the Green-Lagrange strain tensor
+     *             // associated with the I'th vector-valued basis function
```

Comment:
above you've been pretty consistent with ending sentences with a period, except here and a couple of lines below